Support explicit heading IDs [Fixes #272]

think-in-universe opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Update the application to support consistent anchor links across languages.

Tasks:
- [x] Add support for custom ID markdown syntax (like reactjs/reactjs.org#1636)
- [x] Create a script to generate custom heading ID for all the languages

Meet Submission Requirements:
- [x] All anchor links of all languages are the English version (e.g. the German link for the "What is ETH, and how do I get it?" section should be https://ethereum.org/de/use/#_2-what-is-eth-and-how-do-i-get-it, not what it is currently: https://ethereum.org/de/use/#_2-was-ist-eth-und-wie-bekomme-ich-es)
- [x] The solution is flexible enough to automatically support future translations & header links. The current solution supports the heading anchor syntax of  `## Heading {#head1}`, which should be quite easy to follow by translators. 

## Related Issue

Resolves: #272

## Screenshots:

As shown below, the same heading anchor is applied for all the languages: `developers/#best-practices-patterns-and-anti-patterns`

English:

![image](https://user-images.githubusercontent.com/46699230/67889144-44826b80-fb89-11e9-8a5e-1a055fd36d95.png)


Deutsch:

![image](https://user-images.githubusercontent.com/46699230/67889109-36cce600-fb89-11e9-9651-a32bb65c5de7.png)


Chinese:

![image](https://user-images.githubusercontent.com/46699230/67889012-0be29200-fb89-11e9-996f-2f7968e74b6f.png)


And the illustrated problem described in the submission requirement is already resolved:

>  (e.g. the German link for the "What is ETH, and how do I get it?" section should be https://ethereum.org/de/use/#_2-what-is-eth-and-how-do-i-get-it, not what it is currently: https://ethereum.org/de/use/#_2-was-ist-eth-und-wie-bekomme-ich-es)

![image](https://user-images.githubusercontent.com/46699230/67889336-9a571380-fb89-11e9-8493-4b6b5bd0abbf.png)




## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the contributing guidelines in the project README.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

think-in-universe commented:
@samajammin may need comprehensive review from the community of the multi-language documents, since this change involves impacts the heading anchors of all the languages.

think-in-universe commented:
@samajammin could you please also review this PR? I has also been pending for 4~5 days

samajammin commented:
@think-in-universe excellent work here! This review will take some time but I'll try to wrap it up tomorrow.

I have a question on your script (scripts/generateHeadingIDs.js). You described the purpose is:
> to generate custom heading ID for all the languages

Could you please explain this in a bit more depth? At a quick look through, it appears it gathers the header IDs of the English files into a `toc` object, then looks to apply those same header IDs (at the corresponding headers) for all other languages (if needed). From testing out the script a few times it appears the `toc` may not be getting build correctly.

Here's the output from pulling down this PR w/ a fresh yarn install & running `yarn generate:heading-ids`:
https://gist.github.com/samajammin/b0933c54f0a09c1561026b762eb02f8d

After removing a few header IDs (e.g. from /docs/de/developers/index.md) & re-running the script, nothing changed with any files. I expected those missing headers to be added. Here's the output:
https://gist.github.com/samajammin/4597a5a894223ab082e2f1bd3b784906

Please let me know if I'm misunderstanding or if there is in fact a bug.

think-in-universe commented:
@samajammin this is a good question. 

There're several comments worth mentioning:
1. yes. the log does look confusing, and I have updated the log in this commit https://github.com/ethereum/ethereum-org-website/pull/369/commits/dc674ec376e9c7d77cd015a783f420c63c4345ca, and the log should look like this gist: https://gist.github.com/think-in-universe/253a328b4b389ba397239f202c64feb1.   The previous version didn't ignore the case when the target file's header is already processed. In this gist, you could see that a lot of level 3 headings in non-English `developers` doc are not generated, because the level 3 headings don't exist in the English version, (which are formatted in **bold text** instead). So we would either add level 3 headings for English version, or remove the level 3 headings in non-English docs. 
1. the script is written to run against the clean version of the documents that contains no explicit heading IDs. But it should be good to run it incrementally as you did here (because the docs are already processed in the commits). As a result, if you add the level 3 headings for `developers` doc in the English version, the corresponding headings in non-English version should be generated. 
1. It's also worth noticing that the headings are not aligned exactly in some docs between English and non-English docs, such as `learn` (no Enterprise Ethereum section in non-English version except fa), and `developers`. I fixed that by doing some manual change to the English doc to align the headings, so the script could work.

samajammin commented:
Super helpful explanation! Thank you.

Regarding your point here:

> 3. It's also worth noticing that the script is designed by following the assumption that the headings structure are aligned between English and non-English docs (i.e. they share the same ToC -- table of content). 

As you pointed out, this is a risky assumption. **While our goal is to have the translated pages share the same content (& header structure) as the English version, this will not consistently be the case**. We expect the translations to lag behind the English versions (such as the missing `Enterprise Ethereum` section of /learn - eventually the other languages will have it but currently they do not). 

As to your last point:
> the script was not designed to be an actually reusable command since I thought this update of the heading anchors is an one-time effort

I think that's fair. Given that the header structure of translated file is not guaranteed to match the corresponding English header structure, it would be difficult to solve this programmatically.

I believe a simpler script that would be consistently useful for this translation process is a script that, when given a filename, iterates through all headers of the file & generates explicit header IDs if they don't yet exist. An example use case I'm thinking of here is when we want to translate the /docs/enterprise/index.md page ([ethereum.org/enterprise](https://ethereum.org/enterprise/)). We'll want the English page to have those explicit IDs so that the translated files also have them (when new English files are added to CrowdIn, they begin as copies of the English page). So e.g. by running:
```
yarn generate-header-ids docs/enterprise/index.md
```


This script isn't a specific requirement of the bounty & therefore is not necessary but we'd welcome the contribution 😄 

Thanks again for this submission! Excellent work.

samajammin commented on docs/de/developers/index.md:
```diff
@@ -49,16 +49,16 @@ Jedes Programm, das auf der Ethereum Virtual Machine (EVM) läuft, wird allgemei
 Ethereum verfügt über eine große und wachsende Anzahl von Tools, um Programmierern beim Entwickeln, Testen und Bereitstellen von Anwendungen zu helfen. Nachfolgend sind die beliebtesten Tools aufgeführt, mit denen du direkt loslegen kannst. Wenn du tiefer eintauchen möchtest, schau dir diese [umfassende Liste](https://github.com/ConsenSys/ethereum-developer-tools-list) an.
 
-### Truffle: *Entwicklungsumgebung, Test-Framework, Build-Pipeline und weitere Tools*
+### Truffle: *Entwicklungsumgebung, Test-Framework, Build-Pipeline und weitere Tools* {#frameworks}
```
Doesn't look like these h3's match up 😕

think-in-universe replied:
Yeah. We could either remove all the h3 heading IDs in non-English versions, or make changes later.

think-in-universe commented:
@samajammin sounds good. I think that's a quick win, and I have updated the script so you'll be able to run the script against an folder, e.g.

```bash
yarn run generate-heading-ids docs/enterprise
```

and the custom heading IDs of the markdown files under that folder will be added. The log is added in the gist: https://gist.github.com/think-in-universe/a979a3d12e97b3949ac72e78c4462587

I didn't commit the update to `docs/enterprise/index.md` so you could have a try in your own environment and  add into commit if needed.

think-in-universe commented:
@samajammin the anchor links to the English version should have been fixed now.  https://deploy-preview-369--ethereumorg.netlify.com/de/use/

I modified `generateHeadingIDs` script a bit to help speed up the anchor links update.

think-in-universe commented on docs/.vuepress/theme/utils/markdown.js:
```diff
@@ -0,0 +1,72 @@
+      .forEach(token => {
+        token.content = token.content.replace(match[1], '')
+      })
+    // console.log("title", getTitle(textToken)); // test whether the title is updated
```
The comment should have been removed.

think-in-universe commented:
ToDos:

- [ ] create a markdown-it plugin to support the **heading anchor** syntax.
